### ISSUE TITLE
Eta-equivalence in `assert`?

### ISSUE THREAD

**Issue by winitzki** [dhall-lang/dhall-lang#1376]
The standard says in https://github.com/dhall-lang/dhall-lang/blob/master/standard/equivalence.md

> Note that this definition of equivalence does not include η-equivalence, so λ(f : Bool → Bool) → λ(x : Bool) → f x and λ(f : Bool → Bool) → f are not equivalent.

But the recent versions of Dhall actually think they are equivalent:

```dhall
$ dhall repl
Welcome to the Dhall v1.42.1 REPL! Type :help for more information.
⊢ assert : (λ(f : Bool → Bool) → λ(x : Bool) → f x) ===  (λ(f : Bool → Bool) → f)

assert : (λ(f : Bool → Bool) → λ(x : Bool) → f x) ≡ (λ(f : Bool → Bool) → f)
```

Eta-equivalence also works under lambda:

```dhall
⊢ λ(f : Bool → Bool) → assert : f === ( λ(x : Bool) → f x )

λ(f : Bool → Bool) → assert : f ≡ (λ(x : Bool) → f x)
```

(Parentheses around `(λ(x : Bool) → f x)` are required, or else parsing fails.)

If you ask me, of course I'm 100% in favor of using eta-equivalence everywhere. There isn't any difference between `f` and `λx → f x` in a pure language.

Also, it's a great tool to verify laws symbolically. For example, we can define the function composition operation and then verify its associativity law:

```dhall
let compose
  : ∀(a : Type) → ∀(b : Type) → ∀(c : Type) → (b → c) → (a → b) → (a → c)
  = λ(a : Type) → λ(b : Type) → λ(c : Type) → λ(f : b → c) → λ(g : a → b) → λ(x : a) →
    f (g (x)) 
in                -- Let's verify the associativity law. 
   λ(a : Type) → λ(b : Type) →  λ(c : Type) → λ(d : Type) →
      λ(f : a → b) → λ(g : b → c) → λ(h : c → d) →
          assert : 
              compose a b d (compose_backward b c d h g) f
              ≡ compose a c d h (compose_backward a b c g f)
```

But this goes explicitly against the standard.

Also I'd like to see how this is actually implemented in the Haskell version, and I'd like to see the corresponding judgments written into the standard.

There don't seem to be any tests that exercise this behavior. My Scala implementation follows the standard to the letter but fails to type-check `assert : f === (λ(x : Bool) → f x)`, while the Haskell version accepts that.

**Comment by Gabriella439** [comment]
Oh, interesting.  I think I understand why the Haskell implementation is (unintentionally) doing that: it's because in the normalization-by-evaluation logic it's η-expanding all functions.  I'd be okay with a change to standardize η-equivalence.

**Comment by winitzki** [comment]
Would it be sufficient to add to the rules for beta-reduction a new rule that any term `λ(x : a) → f x` is always reduced to just `f`?
Would it be a breaking change because it changes some normal forms? For instance, would some SHA256 values change in the Prelude after this?

**Comment by MonoidMusician** [comment]
Yes, that beta reduction should work. To formulate it properly you need to make sure that the variable `x@0` does not occur in the term `f`, otherwise shifting `x@0` out of `f` for the reduction will not work.

e.g. `\(x : a) -> g x@1 x` needs to go to `g x` and `\(x : a) -> g x x` needs to not eta-reduce

**Comment by winitzki** [comment]
It seems from @Gabriella439 's description that the Haskell version is eta-expanding the functions, not eta-reducing them.

So, should the standard be revised to say that the normal form of a term `f` is `λ(x : Bool) → f x` when `f` is a function type? (The shifting of `x` inside `f` needs to be done carefully, as @MonoidMusician explained.) 

Or should we say that the normal form of `f` is just `f` and the normal form of `λ(x : Bool) → f x` is also `f` (again, with suitable caveats in case `f` contains a free occurrence of `x`)?

**Comment by MonoidMusician** [comment]
Eta expansion and eta reduction should give the same notion of equivalence. Since Dhall doesn't have type-directed normalization, you have to use eta-reduction for the normalization algorithm.

**Comment by winitzki** [comment]
I'm all for using eta-reduction. 

However, this is likely a breaking change. I have implemented additional reduction rules along the lines of what we discussed here. After that, some of the standard tests started failing. A significant number of tests says that the normal form of something must be equal to an eta-expanded form rather than to an eta-reduced form. For example, the test `ListHeadEmptyA.dhall` has the old normal form `λ(T : Type) → None T` which now will become just `None`. The test `MergeA.dhall` expects the normal form `λ(T : Type) → λ(f : ∀(_ : T) → Bool) → λ(x : T) → f x` which will now become the eta-reduced normal form ` λ(T : Type) → λ(f : ∀(_ : T) → Bool) → f`.

I also expect that a few of the hash values from the library will change, as we are changing the definition of the normal form.

**Comment by Gabriella439** [comment]
Yeah, I think it's fine to do this just for `≡`

**Comment by winitzki** [comment]
@Gabriella439 I wonder what was originally the intention of the `assert` feature in Dhall?  It is a feature that one can use only in certain limited ways (for example, no run-time checks and no dependent type checks) and yet it is a powerful feature because `assert` is able to verify symbolic equations under lambda (as in the example with the associativity law of composition that I have shown above).

I have been experimenting with `assert` to verify various things symbolically. The fact that the Haskell implementation accidentally applies eta-normalization means that Dhall can already verify a number of laws symbolically, for example, the laws of monoids or the laws of monads. This could mean potentially to `assert` symbolically that some business logic is correct. Is this an interesting use case? 

In order to be able to verify a wider range of laws, the following features would need to be added as extra steps during the equivalence check:

- Rewrite `a + (b + c)` into `(a + b) + c` and so on, for all Dhall operators (which are all associative).
- Rewrite no-op record operations: for example, rewrite`{ x = r.x, y = r.y}` into just `r` when `r` has a record type with fields `x`, `y`.
- Rewrite `merge a (merge b x)` into a single `merge c x` where `c` is a composition of handlers from `a` and `b`.

There might be other such symbolic simplifications; this is what I have found so far.

What about rewriting `x + 1 + 1` into `x + 2`? Probably that's too much to require of Dhall (a full support for theorem proving in arithmetic would be needed?).

**Comment by Gabriella439** [comment]
So I actually put up a proposal a while ago for Dhall to re-associate associative operators just like you suggested:

https://github.com/dhall-lang/dhall-lang/pull/274

It was rejected at the time (because some other Dhall implementations felt that it increased the complexity of the standard too much) but I'm still in favor of the idea (and those maintainers are no longer participating in Dhall development as far as I can tell).

The second rewrite (for records) might be trickier because it requires knowing the record type at normalization time, but if you know a good way to do this then I'd be okay with it.

I'm also okay with the idea to combine two `merge`s, too.

**Comment by winitzki** [comment]
In your old proposal, how did you achieve rewriting `1 + ((2 + (x + 4)) + 5)` into `3 + x + 9`? This does not simply follow from rewriting all associative operations into left-associative forms. The left-associative form of `1 + ((2 + (x + 4)) + 5)` is `((3 + x) + 4) + 5`.

**Comment by Gabriella439** [comment]
oh, i don't remember why that worked

**Comment by winitzki** [comment]
> oh, i don't remember why that worked

I looked at your old PR, - you suggested to duplicate all rules for associative operators associated to the left or to the right, whenever literal values were involved. This duplicates the number of rules. I'm just proposing to add a rewrite rule to make all operator expressions always left-associated, so that normal forms are somewhat more canonical (but not fully canonical).
